libbpf-tools/doublefree.c

```diff
@@ -0,0 +1,473 @@
+/* SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause) */
+/* Copyright (c) 2022 LG Electronics */
+
+// 19-Oct-2022   Bojun Seo   Created this.
+#include <argp.h>
+#include <stdio.h>
+#include <signal.h>
+#include <stdarg.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include "doublefree.h"
+#include "doublefree.skel.h"
+#include "trace_helpers.h"
+#include "uprobe_helpers.h"
+
+#define STACK_MAX 127
+#define BUF_MAX (STACK_MAX * LINE_MAX * 2)
+#define PERF_BUFFER_PAGES 16
+#define PERF_POLL_TIMEOUT_MS 10000
+#define DATE_FORMAT "%2d-%s-%02d %02d:%02d:%02d "
+
+#define ON_MEM_FAILURE(buf) \
+	do { \
+		if (NULL == buf) { \
+			p_err("FATAL: Failed to allocate memory on %s", __func__); \
+			exit(-1); \
+		} \
+	} while (false)
+
+#define _CHECK_OFFSET(func_off) \
+	do { \
+		if (func_off < 0) { \
+			p_warn("Failed to get func_offset"); \
+			return func_off; \
+		} \
+	} while (false)
+
+#define _CHECK_PROGRAM(obj, func_name) \
+	do { \
+		int err = libbpf_get_error(obj->links.func_name); \
+		if (err) { \
+			p_warn("Failed to attach %s: %d", #func_name, err); \
+			return err; \
+		} \
+	} while (false)
+
+#define _ATTACH_UPROBE(obj, lib_path, func_name, is_uretprobe, func_off, pid) \
+	do { \
+		obj->links.func_name =	bpf_program__attach_uprobe( \
+			obj->progs.func_name, is_uretprobe, pid, lib_path, func_off); \
+	} while (false)
+
+#define ATTACH_UPROBE(obj, lib_path, func_name, pid) \
+	do { \
+		off_t func_off = get_elf_func_offset(lib_path, #func_name); \
+		_ATTACH_UPROBE(obj, lib_path, func_name##_entry, false, func_off, pid); \
+	} while (false)
+
+#define ATTACH_URETPROBE(obj, lib_path, func_name, pid) \
+	do { \
+		off_t func_off = get_elf_func_offset(lib_path, #func_name); \
+		_ATTACH_UPROBE(obj, lib_path, func_name##_return, true, func_off, pid); \
+	} while (false)
+
+#define ATTACH_UPROBES(obj, lib_path, func_name, pid) \
+	do { \
+		off_t func_off = get_elf_func_offset(lib_path, #func_name); \
+		_ATTACH_UPROBE(obj, lib_path, func_name##_entry, false, func_off, pid); \
+		_ATTACH_UPROBE(obj, lib_path, func_name##_return, true, func_off, pid); \
+	} while (false)
+
+#define ATTACH_UPROBE_CHECKED(obj, lib_path, func_name, pid) \
+	do { \
+		off_t func_off = get_elf_func_offset(lib_path, #func_name); \
+		_CHECK_OFFSET(func_off); \
+		_ATTACH_UPROBE(obj, lib_path, func_name##_entry, false, func_off, pid); \
+		_CHECK_PROGRAM(obj, func_name##_entry); \
+	} while (false)
+
+#define ATTACH_URETPROBE_CHECKED(obj, lib_path, func_name, pid) \
+	do { \
+		off_t func_off = get_elf_func_offset(lib_path, #func_name); \
+		_CHECK_OFFSET(func_off); \
+		_ATTACH_UPROBE(obj, lib_path, func_name##_return, true, func_off, pid); \
+		_CHECK_PROGRAM(obj, func_name##_return); \
+	} while (false)
```

Comment:
Sounds great! I always appreciate your valuable opinion.
I prefer the second code you suggested(using struct).
And I think it would be better to pass `struct probe` instance(using pointer) instead of each members to `attach_uprobe` function, like followings.
```C
for (int i = 0; i < sizeof(probes) / sizeof(probes[0]); i++) {
	attach_uprobe(&probes[i], libc_path);
}
```
What you think?

Reply:
I changed uprobe attaching codes. I tried implementing it while incorporating your opinion, but in a slightly different way.